compiler-plugin/src/main/kotlin/arrow/meta/quotes/ClassOrObject.kt

```diff
@@ -13,12 +13,111 @@ import org.jetbrains.kotlin.psi.psiUtil.getValueParameters
+ *        ktClass.typeParameters.isEmpty()
+ * ```
+ *
+ * For a [KtClass] to be a product type, we check to see if the [KtClass] is an ADT, and that the cardinality of
+ * the constructor parameters is greater than one and non-mutable. For the [LensPlugin], we've just demonstrated how we
+ * can check for the product type of a [KtClass] by access the properties made available by the Kotlin PSI change.
+ *
+ * While it is not necessary to build [KtClass] predicates for [Meta.classOrObject] does not necessarily have to serve
+ * as a check for functional proofs in your compiler plugin.  It is relevant to highlight that the `match` predicate
+ * written gives access to the Kotlin PSI when analyzing written code via the written compiler plugin.
+ *
+ * The second parameter `map` is a function that allows the resulting action from matching on the transformation at
+ * the PSI level. Let us example another example plugin:
+ *
+ * ```kotlin:ank:silent
+ * import arrow.meta.Meta
+ * import arrow.meta.Plugin
+ * import arrow.meta.invoke
+ * import arrow.meta.quotes.Transform
+ * import arrow.meta.quotes.classOrObject
+ *
+ * val Meta.example: Plugin
+ *   get() =
+ *     "Example" {
+ *       meta(
+ *         /** Intercepts all classes named 'Test' **/
+ *         classOrObject({ name == "Test" }) { classOrObject: KtClass ->
+ *           Transform.replace(
+ *             replacing = classOrObject,
+ *             newDeclaration =
+ *               """|$`@annotationEntries` $kind $name $`(typeParameters)` $`(valueParameters)` : $supertypes"} {
+ *                  |  $body
+ *                  |  fun void test(): Unit =
+ *                  |    println("Implemented by ΛRROW Meta!")
+ *                  |}
+ *                  |""".`class`.synthetic
+ *           )
+ *         }
+ *       )
+ *     }
+ * ```
+ *
+ * After analyzing the PSI elements available, we pass a resulting [KtClass] matching the predicate (in our case,
+ * we pass the resulting [KtClass] whose name is "Test") and replace the entire object with the string block, which is
+ * then wrapped as a [ClassScope] and wrapped so the `newDeclaration` is of the type [Scope]<[ClassScope]> to match
+ * match compatibility of the intercepted classes wrapped in some kind of [Scope]. Too see more, scroll down to
+ * [ClassScope].
+ *
+ * @param match designed to to feed in any kind of [KtClass] predicate returning a [Boolean]
+ * @param map
+ */
 fun Meta.classOrObject(
   match: KtClass.() -> Boolean,
   map: ClassScope.(KtClass) -> Transform<KtClass>
 ): ExtensionPhase =
   quote(match, map) { ClassScope(it) }
 
+/**
+ * The [ClassScope] is projected over the template to allow destructuring of the different parts of the
+ * [KtClass]. The scope enables template syntax where the user may add new members or modify the class structure
+ * before it's compiled.
+ *
+ * @param value the PSI element being fed in for the compiler plugin.
+ * @param `@annotationEntries`
+ * @param modality
```

Comment:
KDocs doesn't appear to generate `quotes` as parameter names here smh

<img width="952" alt="Screen Shot 2019-11-06 at 4 29 35 PM" src="https://user-images.githubusercontent.com/6917498/68343697-fa324a80-00b2-11ea-9869-b5da624468bf.png">

Reply:
Do you mean "param `@annotationEntries\`"?
I am not sure if Dokka supports that properly.

Reply:
I can just go with annotationEntries perhaps to keep the same formatting if that makes more sense?

Reply:
Also, what is the significance that we want this parameter to be read as a string? I thought it was curious but it might be worth commenting on in the documentation, regardless whether it's significant or nah